bot/utils/redis_cache.py

```diff
@@ -0,0 +1,258 @@
+        self.bot = None
+
+    def _set_namespace(self, namespace: str) -> None:
+        """Try to set the namespace, but do not permit collisions."""
+        while namespace in self._namespaces:
+            namespace += "_"
+
+        self._namespaces.append(namespace)
+        self._namespace = namespace
+
+    @staticmethod
+    def _valid_typestring_types() -> str:
+        """
+        Creates a nice, readable list of valid types for typestrings, useful for error messages.
+
+        This will be dynamically updated if we change the TYPESTRING_PREFIXES constant up top.
+        """
+        valid_types = ", ".join([str(_type).split("'")[1] for _, _type in TYPESTRING_PREFIXES])
+        valid_types = ", and ".join(valid_types.rsplit(", ", 1))
+        return valid_types
+
+    @staticmethod
+    def _valid_typestring_prefixes() -> str:
+        """
+        Creates a nice, readable list of valid prefixes for typestrings, useful for error messages.
+
+        This will be dynamically updated if we change the TYPESTRING_PREFIXES constant up top.
+        """
+        valid_prefixes = ", ".join([f"'{prefix}'" for prefix, _ in TYPESTRING_PREFIXES])
+        valid_prefixes = ", and ".join(valid_prefixes.rsplit(", ", 1))
+        return valid_prefixes
+
+    def _to_typestring(self, value: ValidRedisType) -> str:
+        """Turn a valid Redis type into a typestring."""
+        for prefix, _type in TYPESTRING_PREFIXES:
+            if isinstance(value, _type):
+                return f"{prefix}{value}"
+        raise TypeError(f"RedisCache._from_typestring only supports the types {self._valid_typestring_types()}.")
+
+    def _from_typestring(self, value: Union[bytes, str]) -> ValidRedisType:
+        """Turn a typestring into a valid Redis type."""
+        # Stuff that comes out of Redis will be bytestrings, so let's decode those.
+        if isinstance(value, bytes):
+            value = value.decode('utf-8')
+
+        # Now we convert our unicode string back into the type it originally was.
+        for prefix, _type in TYPESTRING_PREFIXES:
+            if value.startswith(prefix):
+                return _type(value[2:])
+        raise TypeError(f"RedisCache._to_typestring only supports the prefixes {self._valid_typestring_prefixes()}.")
+
+    def _dict_from_typestring(self, dictionary: Dict) -> Dict:
+        """Turns all contents of a dict into valid Redis types."""
+        return {self._from_typestring(key): self._from_typestring(value) for key, value in dictionary.items()}
+
+    def _dict_to_typestring(self, dictionary: Dict) -> Dict:
+        """Turns all contents of a dict into typestrings."""
+        return {self._to_typestring(key): self._to_typestring(value) for key, value in dictionary.items()}
+
+    async def _validate_cache(self) -> None:
+        """Validate that the RedisCache is ready to be used."""
+        if self.bot is None:
+            raise RuntimeError("Critical error: RedisCache has no `Bot` instance.")
```

Comment:
I liked what you did for the namespace error message. Please make this error more specific too. Unless someone reads the class docstring (which we hope they do, but no guarantees), they won't know what this means/how to fix it.